src/lib/core/diag.c

```diff
@@ -53,6 +53,7 @@ error_unref(struct error *e)
 			to_delete->cause->effect = NULL;
 			to_delete->cause = NULL;
 		}
+		error_payload_destroy(&to_delete->payload);
 		to_delete->destroy(to_delete);
```

Comment:
Good point. The problem of `destroy()` is that it just always calls `free()` inside. We do not call `free()` right away because the errors are C++ objects and we can't call a C++ destructor in C except when it is hidden inside a function stored by plain pointer like `error.destroy`.

It should be gone when we drop C++ from errors code. Until then not sure what to do here. If the destructor will change before we delete C++ and will use the payload, then at least we will get an early crash because `error_payload_destroy` calls `TRASH` on self.

If I would try to destroy the payload after `error->destroy(error)` it would crash - the memory is freed already. If I would destroy it before, but keep it usable (not call `TRASH` on it), and `error->destroy` would add new fields - they would leak, because after this call I can't destroy the payload again - the memory is freed.

A solution would be to split it into separate `error->destroy` and `error->free` calls, but I would rather wait until we just drop C++. Sounds good?

Reply:
> `to_delete->destroy` represents derived class destructor and thus must be called before destruction of struct error. At least because it can use this payload for some reason.

IMO we should design our callbacks as we like so long as the code is properly documented and easy to read. There's no need to follow any paradigm just for the sake of it. Let's deal with this problem only when (if) a derived class destructor actually *needs* to access the payload (which will probably never happen).

Reply:
Ahh, `->destroy` actually deletes the object.
I don't propose C++ paradigm because C++ is good, I propose it because it's correct and safe in general. An even if we drop C++ completely, we should name the method that deletes object as `_delete` and think about each object's lifetime.
I would definitely prefer `error->destroy` and `error->free` pair, but I understand that's out of scope of the issue, so up to you to decide.

Reply:
`delete` is a reserved key word in C so using it would be inconvenient. I don't see anything wrong for the `destroy` callback to free the underlying object. Splitting `destroy` to `destroy` and `free` would be an overkill IMO.

Reply:
In C we usually combine name of function from name of object and name of method, so we don't interfere with `delete` keyword.
Let's just do according to our code guidelines: https://www.tarantool.io/en/doc/latest/dev_guide/c_style_guide/#chapter-4-naming

Reply:
> In C we usually combine name of function from name of object and name of method,

But when it's a callback, we don't: we use `obj->vtab.destroy`, not `obj->vtab.obj_destroy`.

> so we don't interfere with `delete` keyword. And I bet we don't need `delete` member here, just `destroy` and `free` pair, a likely `error_delete` function.

There's no need to introduce `destroy`/`free` pair - everything works just fine with one callback; no need to add yet another.

> Let's just do according to our code guidelines: https://www.tarantool.io/en/doc/latest/dev_guide/c_style_guide/#chapter-4-naming

It says nothing about callbacks. We can call a constructor/destructor pair `obj_new`/`obj_delete`, but under the hood call the `destroy` callback, which would free the object.